tokio-quiche/src/quic/router/multipath.rs

```diff
@@ -0,0 +1,684 @@
+//
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+// IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+// THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use std::future::Future;
+use std::io;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::Context;
+use std::task::Poll;
+
+use datagram_socket::DatagramSocketRecv;
+use datagram_socket::DatagramSocketSend;
+use task_killswitch::spawn_with_killswitch;
+use tokio::sync::mpsc;
+
+use crate::buf_factory::BufFactory;
+use crate::buf_factory::PooledBuf;
+use crate::metrics::labels;
+use crate::metrics::quic_expensive_metrics_ip_reduce;
+use crate::metrics::Metrics;
+use crate::quic::connection::InitialQuicConnection;
+use crate::quic::router::initial_packet_error_type;
+use crate::quic::router::short_dcid;
+use crate::quic::router::ConnStream;
+use crate::quic::router::ConnectionMap;
+use crate::quic::router::ConnectionMapCommand;
+use crate::quic::router::InitialPacketHandler;
+use crate::quic::router::PollRecvData;
+use crate::quic::Incoming;
+use crate::settings::Config;
+
+/// A router that can handle incoming packets from multiple network paths.
+pub struct MultiPathInboundRouter<Tx, Rx, M, I>
+where
+    Tx: DatagramSocketSend + Send + 'static,
+    M: Metrics,
+{
+    // Shared connection map
+    conns: ConnectionMap,
+
+    // Path information
+    path_rxs: Vec<(Rx, std::net::SocketAddr)>, // (receiver, local_addr) pairs
+    path_txs: Vec<Arc<Tx>>,                    // Shared senders for all paths
+
+    // Primary path info (for backward compatibility)
+    primary_local_addr: std::net::SocketAddr,
+
+    // Connection management
+    incoming_packet_handler: I,
+    shutdown_tx: Option<mpsc::Sender<()>>,
+    shutdown_rx: mpsc::Receiver<()>,
+    conn_map_cmd_tx: mpsc::UnboundedSender<ConnectionMapCommand>,
+    conn_map_cmd_rx: mpsc::UnboundedReceiver<ConnectionMapCommand>,
+    accept_sink: mpsc::Sender<io::Result<InitialQuicConnection<Tx, M>>>,
+
+    // Configuration and metrics
+    config: Config,
+    metrics: M,
+
+    // Buffer management
+    buffers: Vec<PooledBuf>, // One buffer per path
```

Comment:
I see that there is one buffer per path (which is fine), but it is not clear how the link is made between the PathId and its buffer. Is the PathId the index? In such case, it may cause troubles in long connections having many path abandons and path creations.